src/PublicSuffixListManager.php

```diff
@@ -57,68 +48,82 @@ class PublicSuffixListManager
     /**
      * Public constructor.
      *
-     * @param HttpAdapter $httpAdapter
-     * @param string      $cacheDir    Optional cache directory
+     * @param CacheInterface $cacheAdapter
+     * @param HttpAdapter    $httpAdapter
+     * @param string         $sourceUrl
      */
-    public function __construct(HttpAdapter $httpAdapter, string $cacheDir = null)
-    {
-        $this->cacheDir = $cacheDir ?? realpath(dirname(__DIR__) . DIRECTORY_SEPARATOR . 'data');
+    public function __construct(
+        CacheInterface $cacheAdapter,
+        HttpAdapter $httpAdapter,
+        string $sourceUrl = self::PUBLIC_SUFFIX_LIST_URL
+    ) {
+        $this->cacheAdapter = $cacheAdapter;
         $this->httpAdapter = $httpAdapter;
+        $this->sourceUrl = $sourceUrl;
     }
 
     /**
      * Gets Public Suffix List.
      *
-     * @param string $list the Public Suffix List type
+     * @param string $type the Public Suffix List type
+     *
+     * @throws RuntimeException if no PublicSuffixList can be returned
      *
      * @return PublicSuffixList
      */
-    public function getList($list = self::ALL_DOMAINS): PublicSuffixList
+    public function getList($type = self::ALL_DOMAINS): PublicSuffixList
     {
-        $cacheBasename = isset(self::$domainList[$list]) ? self::$domainList[$list] : self::PDP_PSL_PHP_FILE;
-        $cacheFile = $this->cacheDir . '/' . $cacheBasename;
-        if (!file_exists($cacheFile)) {
-            $this->refreshPublicSuffixList();
+        static $type_lists = [self::ALL_DOMAINS => 1, self::ICANN_DOMAINS => 1, self::PRIVATE_DOMAINS => 1];
+
+        if (!isset($type_lists[$type])) {
+            $type = self::ALL_DOMAINS;
+        }
+
+        if (($list = $this->cacheAdapter->get($type)) !== null) {
+            return new PublicSuffixList($list);
+        }
+
+        if (($raw = $this->cacheAdapter->get(self::PUBLIC_SUFFIX_LIST_DATA)) !== null
+            && $this->cachePublicSuffixListTypes($raw)) {
+            return new PublicSuffixList($this->cacheAdapter->get($type));
         }
 
-        return new PublicSuffixList($cacheFile);
+        if ($this->refreshPublicSuffixList()) {
+            return new PublicSuffixList($this->cacheAdapter->get($type));
+        }
+
+        //this should not happen but in rare cases you should be able to cach this one
+        throw new RuntimeException(sprintf('Unable to return `%s` Public Suffix List', $type));
```

Comment:
I used a RuntimeException here but if we want to be nice to people we could create a base Exception for the library and just use this exception so that users can detect the exception origin more easily